### PR TITLE
Modify Side Input Data Types for cuDNN BatchNorm

### PR DESCRIPTION
**Pull request by AyanmoI**
Side inputs for cuDNN Fused BatchnormForward have been taking the incorrect data type till now. It should be the same data type as the input. This PR aims to correct that in the `stream_executor` and `core/kernels`.

### CODE REVIEW

**Review by sanjoy — CHANGES_REQUESTED** [review]
Ayan, is it possible to write a unit test for this?

**Comment by gbaned** [comment]
@AyanmoI  Can you please check @sanjoy's comments and keep us posted ? Thanks!

**Comment by kaixih** [comment]
Hi @sanjoy, I think this is an obvious wrong use of data type of side input. As shown below, the side_input should be T but we convert it later to U:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/ops/nn_ops.cc#L228
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/fused_batch_norm_op.cc#L913.

Unfortunately or fortunately, it is hard to write a unit test case to hit it, because we've limit the U to be float, which is wider than or equal to the T (half or float). So, always using U/float for the side_input would be actually "safe" enough when we use the AsDeviceMemory<U> to convert a half tensor to obtain its raw pointer. However, this bug can be encountered when we want to explicitly call the `ThenBatchNormalizationForward` as shown in tensorflow/compiler/xla/service/gpu/cudnn_batchnorm_runner.cc from this PR.

**Review by sanjoy — APPROVED** [review]
